### PR TITLE
Add errors for sequence insert or update.

### PR DESCRIPTION
**Pull request by mandyvenables**

### CODE REVIEW

File: lib/storage.js [hunk 1]
```diff
@@ -286,6 +286,9 @@ api.insert = async ({edvId, doc}) => {
   if(doc.sequence < 0) {
     throw new TypeError('"doc.sequence" must be a non-negative integer.');
   }
+  if(doc.sequence >= Number.MAX_SAFE_INTEGER) {
```
Comment (mattcollier) [inline lib/storage.js:289]: I think you'll want to use `isSafeInteger` for this determiniation.  https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isSafeInteger

Javascript effectively loses its mind if a value exceed MAX_SAFE_INTEGER so  I don't think it can reliably make comparisons and reason about those values.

Under the covers I suspect this helper is something like this which tests to ensure that the value is LESS THAN or equal to MAX_SAFE_INTEGER

http://speakingjs.com/es5/ch11.html#safe_integers

in addition to all of that, MAX_SAFE_INTEGER itself should be considered a valid value which you will get for free by using `isSafeInteger`.  Even if the existing code functioned reliably it would be rejecting `doc.sequence = Number.MAX_SAFE_INTEGER`.

File: test/mocha/12-insert.js [hunk 2]
```diff
@@ -95,6 +95,29 @@ describe('insert API', () => {
...
     });
+  it('should fail to insert a document with an unsafe sequence number',
+    async () => {
+      let record, error = null;
+      try {
+        const actor = actors['alpha@example.com'];
+        const {doc1} = mockData;
+        const doc = {...doc1};
+        doc.id = await helpers.generateRandom();
+        doc.sequence = Number.MAX_SAFE_INTEGER;
```
[the remark this answers is not in the document]
Comment (mandyvenables) [reply test/mocha/12-insert.js:106]: It was agreed that MAX_SAFE_INTEGER is not acceptable, so I changed the description of the test.

File: lib/storage.js (hunk 1 above)
Comment (mandyvenables) [reply lib/storage.js]: Per our conversation during office hours, I added in the check for both isSafeInteger and less than MAX_SAFE_INTEGER.

File: test/mocha/helpers.js [hunk 3]
```diff
@@ -57,10 +57,7 @@ exports.sequenceNumberTests = [
   ['in range [2**32, MAX_SAFE_INTEGER] (middle)',
     2 ** 32 + (Number.MAX_SAFE_INTEGER - 2 ** 32 - 1) / 2],
   ['in range [2**32, MAX_SAFE_INTEGER] (random)',
-    getRandomIntInclusive(2 ** 32, Number.MAX_SAFE_INTEGER)],
-  // near Number.MAX_SAFE_INTEGER
-  ['MAX_SAFE_INTEGER-1', Number.MAX_SAFE_INTEGER - 1],
-  ['MAX_SAFE_INTEGER', Number.MAX_SAFE_INTEGER],
+    getRandomIntInclusive(2 ** 32, Number.MAX_SAFE_INTEGER)]
```
[the remark this answers is not in the document]
Comment (mandyvenables) [reply test/mocha/helpers.js]: I updated the test to be MAX_SAFE_INTEGER - 1 so the test will not fail on the occasion that the random number is MAX_SAFE_INTEGER.

File: lib/storage.js [hunk 4]
```diff
@@ -286,6 +286,10 @@ api.insert = async ({edvId, doc}) => {
   if(doc.sequence < 0) {
     throw new TypeError('"doc.sequence" must be a non-negative integer.');
   }
+  if(!Number.isSafeInteger(doc.sequence) ||
```
Comment (mattcollier) [inline lib/storage.js:289]: looks like there's an opportunity to DRY this up a bit.  How about a private helper `_validateDocSequence(..)` helper used in both places?

Also, technically the value has not exceeded MAX_SAFE_INTEGER so I wonder if that's going to be confusing. Not sure what to say here that lets people know what the safe values are unless its exacly `exceeded MAX_SAFE_INTEGER - 1`.  @mandyvenables @dlongley thoughts, suggestions?

File: test/mocha/12-insert.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (mattcollier) [reply test/mocha/12-insert.js:106]: can you please add comments here and/or in the code that explains the rational behind the choice we made today?  Without all the context that we went over on the call todday, I think it would be difficult for someone to determine why MAX_SAFE_INTEGER is not acceptable.  Something like:

doc.sequence is limited to MAX_SAFE_INTEGER - 1 to avoid unexpected behavior when a client attempts to increment the sequence number.

File: lib/storage.js (hunk 4 above)
Comment (davidlehn) [reply lib/storage.js:289]: In my earlier comment I suggested just saying the sequence number is too large.

File: test/mocha/helpers.js [hunk 5]
```diff
@@ -57,10 +57,7 @@ exports.sequenceNumberTests = [
   ['in range [2**32, MAX_SAFE_INTEGER] (middle)',
     2 ** 32 + (Number.MAX_SAFE_INTEGER - 2 ** 32 - 1) / 2],
   ['in range [2**32, MAX_SAFE_INTEGER] (random)',
-    getRandomIntInclusive(2 ** 32, Number.MAX_SAFE_INTEGER)],
-  // near Number.MAX_SAFE_INTEGER
-  ['MAX_SAFE_INTEGER-1', Number.MAX_SAFE_INTEGER - 1],
```
Comment (davidlehn) [inline test/mocha/helpers.js:62]: Should add back the MAX - 1 test since that's the edge case now.

Comment (mandyvenables) [reply test/mocha/helpers.js:62]: We are using this same export to also test the update which increases the sequence by 1. I changed it to MAX_SAFE_INTEGER - 2 so it will test the ceiling on update.

File: test/mocha/12-insert.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (mandyvenables) [reply test/mocha/12-insert.js:106]: I added the reasoning comment in the _validateDocSequence function in the storage.js file.

File: lib/storage.js (hunk 4 above)
Comment (mandyvenables) [reply lib/storage.js]: Moved the check to a private helper function and updated the error message to '"doc.sequence" number is too large.'

File: lib/storage.js [hunk 6]
```diff
@@ -666,3 +668,12 @@ function _assert128BitId(id) {
       'array of 16 random bytes.');
   }
 }
+
+function _validateDocSequence(sequence) {
```
Comment (davidlehn) [inline lib/storage.js:672]: Might be good to also move the "seq < 0" check here to avoid repeating it twice above.

Comment (mandyvenables) [reply lib/storage.js:661]: @davidlehn Done, the non-negative sequence check is now in the private helper function.
